File: core/commands/registry/irules/pem__flow.py
# Enriched from F5 iRules reference documentation.
"""PEM::flow -- PEM iRule command for flow features, including transacitonal and eval."""

from __future__ import annotations

from ....compiler.side_effects import ConnectionSide, SideEffect, SideEffectTarget
from .._base import CommandDef
from ..models import CommandSpec, FormKind, FormSpec, HoverSnippet, ValidationSpec
from ..namespace_models import EventRequires
from ..signatures import Arity
from ._base import _IRULES_ONLY, register

_SOURCE = "https://clouddocs.f5.com/api/irules/PEM__flow.html"


@register
class PemFlowCommand(CommandDef):
    name = "PEM::flow"

    @classmethod
    def spec(cls) -> CommandSpec:
        return CommandSpec(
            name="PEM::flow",
            dialects=_IRULES_ONLY,
            hover=HoverSnippet(
                summary="PEM iRule command for flow features, including transacitonal and eval.",
                synopsis=(
                    "PEM::flow transactional disable",
                    "PEM::flow eval",
                ),
                snippet=(
                    "The transciontal disable command disables the transactional feature in PEM for a flow.\n"
                    "The eval command trigers the policy evaluation for the flow immediately."
                ),
                source=_SOURCE,
                examples=(
                    "when HTTP_REQUEST {\n"
                    "    PEM::flow transactional disable;\n"
                    "    PEM::flow eval;\n"
                    "}"
                ),
            ),
            forms=(
                FormSpec(
                    kind=FormKind.DEFAULT,
                    synopsis="PEM::flow transactional disable",
                ),
            ),
            validation=ValidationSpec(
                arity=Arity(),
            ),
            event_requires=EventRequires(),
            side_effect_hints=(
                SideEffect(
                    target=SideEffectTarget.CONNECTION_CONTROL,
                    reads=True,
                    connection_side=ConnectionSide.BOTH,
                ),
            ),
        )

File: rust/tcl-registry/src/commands/irules/pem__flow.rs
// tcl-lsp — a language server and toolchain for Tcl
// Copyright (C) 2026 James Deucker (bitwisecook) <https://github.com/bitwisecook>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//
// SPDX-License-Identifier: AGPL-3.0-or-later

//! `PEM::flow` iRules command.
use crate::prelude::*;
pub const fn spec() -> CommandSpec {
    CommandSpec {
        name: "PEM::flow",
        dialects: Some(DialectSet::IRULES),
        arity: Arity::at_least(0),
        hover: Some(HoverSnippet {
            summary: "PEM iRule command for flow features, including transacitonal and eval.",
            synopsis: &["PEM::flow transactional disable", "PEM::flow eval"],
            snippet: "The transciontal disable command disables the transactional feature in PEM for a flow.\nThe eval command trigers the policy evaluation for the flow immediately.",
            source: "https://clouddocs.f5.com/api/irules/PEM__flow.html",
            examples: "when HTTP_REQUEST {\n    PEM::flow transactional disable;\n    PEM::flow eval;\n}",
            return_value: "",
        }),
        forms: &[FormSpec {
            synopsis: "PEM::flow transactional disable",
            ..FormSpec::DEFAULT
        }],
        side_effects: &[SideEffect {
            target: SideEffectTarget::ConnectionControl,
            reads: true,
            connection_side: ConnectionSide::Both,
            ..SideEffect::DEFAULT
        }],
        ..CommandSpec::DEFAULT
    }
}
